tools/slo-generator/slo_generator/report.py

```diff
@@ -0,0 +1,274 @@
+        backend_cfg = {k: v for k, v in cfg.items() if k not in excluded_keys}
+        instance = utils.get_backend_cls(cls)(client=client, **backend_cfg)
+        method = getattr(instance, method)
+        LOGGER.debug(f'{info} | '
+                     f'Using backend {cls}.{method.__name__} (from '
+                     f'SLO config file).')
+
+        # Delete mode activation.
+        if delete and hasattr(instance, 'delete'):
+            method = instance.delete
+            LOGGER.warning(f'{info} | Delete mode enabled.')
+
+        # Run backend method and return results.
+        result = method(self.timestamp, self.window, config)
+        LOGGER.debug(f'{info} | Backend results: {result}')
+        return result
+
+    def get_sli(self, result):
+        """Compute SLI value and good / bad counts from the backend result.
+
+        Some backends (e.g: Prometheus) are computing and returning the SLI
+        value directly, others are sending a tuple (good_count, bad_count) and
+        SLI value is computed from there.
+
+        Args:
+            result (obj): Backend result.
+
+        Returns:
+            tuple: A tuple of 3 values to unpack (float, int, int).
+                float: SLI value.
+                int: Good events count.
+                int: Bad events count.
+
+        Raises:
+            Exception: When the backend does not return a proper result.
+        """
+        info = self.__get_info()
+        if isinstance(result, tuple):
+            good_count, bad_count = result
+            LOGGER.debug(f"{info} | Good: {good_count} | Bad: {bad_count}")
+            if (good_count + bad_count) == 0:
+                LOGGER.error(f"{info} | No events found.")
+                return 0, 0, 0
+            sli_measurement = round(good_count / (good_count + bad_count), 6)
+
+        elif isinstance(result, (float, int)):
+            good_count, bad_count = 0, 0
+            sli_measurement = round(result, 6)
+
+        else:
+            msg = "Backend did not return any valid results."
+            LOGGER.error(msg)
+            LOGGER.debug(f'Backend result: {result}')
+            raise Exception(msg)
+
+        return sli_measurement, good_count, bad_count
+
+    def to_json(self):
+        """Serialize dataclass to JSON."""
+        return asdict(self)
+
+    def is_empty(self):
+        """Return True if report is empty (SLI = 0 or sum of good / bad count
+        is null)."""
+        return self.sli_measurement == 0
+
+    def __set_fields(self, lambdas={}, **kwargs):
+        """Set all fields in dataclasses from configs passed and apply function
+        on values whose key match one in the dictionaries.
+
+        Args:
+            lambdas (dict): Dict {key: function} to apply a function on certain
+            kwargs (dict): Dict of key / values to set in dataclass.
+        """
+        names = set(f.name for f in fields(self))
+        for k, val in kwargs.items():
```

Comment:
We should loop over `names` rather than kwargs, since a non-named kwarg should never be inspected/touched.

Reply:
Renamed `kwargs` to `fields` which seems more appropriate.

Reply:
Good catch, updated to loop through names instead.